### DISCUSSION TITLE
Performance Data implementation

### DISCUSSION THREAD

**Discussion by Byolock** [atc0005/check-vmware#315] [Stable Release]
I've noticed that the plug-ins don't support performance data. Do you plan to implement support?

If not, I'll try implementing it myself I got about a few weeks experience with c#. Doesn't look too bad though if I understand it correctly it's just building a correctly formatted string and append it to the output. I already did some tests with the host cpu check, while debuging the plugin output looks correctly didn't have time yet to compile and test with icinga2.

**Comment by atc0005** [comment]
> @Byolock: I've noticed that the plug-ins don't support performance data. Do you plan to implement support?

I've not yet implemented performance data reporting in this project or in other plugins that I maintain, but I'm open to doing so. I'll need to review some examples and then implement initial support. Are you willing to test prototype versions of the plugins?

> I already did some tests with the host cpu check, while debuging the plugin output looks correctly didn't have time yet to compile and test with icinga2.

Do you happen to have this available in a branch on a fork that I could look at? If not, would you be open to posting your changes here as a diff?

**Reply by Byolock** [reply]
It really is just one metric to check if I would have a chance to teach myself enough go to implement it myself, most likely not pretty and not even at the place it makes the most sense, but here is the fork : https://github.com/Byolock/check-vmware/tree/perfdata

**Reply by atc0005** [reply]
Thanks for sharing, it helps make a bit clearer what is needed.

Some notes for my later review:

- https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/perfdata.html
- https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/pluginapi.html
- https://nagios-plugins.org/doc/guidelines.html#AEN200

I think I'd like to implement this in a small library I use for handling Nagios return codes/output as it feels like it would be a good fit for that package.

**Reply by atc0005** [reply]
> I think I'd like to implement this

To clarify, I was referring to the necessary support for formatting the performance data output.

**Reply by Byolock** [reply]
Thank you for your response. I might have another useful resource for you here :
https://nagios-plugins.org/doc/guidelines.html#AEN200

And to this : 

> Are you willing to test prototype versions of the plugins?

Of course.

**Reply by atc0005** [reply]
Thanks for the link and the willingness to help test. Will let you know once I have something to look at. I'm hoping to carve out some time for this in the next week.

**Reply by atc0005** [reply]
Cross-post (https://github.com/atc0005/check-vmware/discussions/322#discussioncomment-1374533):

> - v0.19.1 is a dependency update only (doesn't include the PerfData output we're testing in v0.20.0-alpha.x releases)
> - [v0.20.0-alpha.3](https://github.com/atc0005/check-vmware/releases/tag/v0.20.0-alpha.3) is the current version available for testing

**Comment by atc0005** [comment]
Recap (for anyone not following existing discussion on #321 and linked threads):

- Initial support for performance data is included in the v0.20.0 release for the `check_vmware_datastore` plugin
- support for additional metrics and updates to other plugins in this project are planned for near future releases

**Reply by atc0005** [reply]
> Recap (for anyone not following existing discussion on #321 and linked threads):
> 
> * Initial support for performance data is included in the v0.20.0 release for the `check_vmware_datastore` plugin
> * support for additional metrics and updates to other plugins in this project are planned for near future releases

See also:

- GH-355
- GH-356
- GH-348
- GH-351
- GH-352
- GH-353
- GH-350
- GH-347
- GH-346
- GH-357
- GH-345
- GH-349
- GH-344

and the related project board here:

- https://github.com/atc0005/check-vmware/projects/1

**Comment by atc0005** [comment]
@Byolock

Current metrics slated for the `check_vmware_host_cpu` plugin per #346:

- `time` (plugin runtime)
- `cpu_usage` (percentage)
- `cpu_total` (Hz)
- `cpu_remaining` (Hz)
- `vms` (total)
- `vms_powered_off`
- `vms_powered_on`

I'm on the fence regarding `cpu_remaining` (measured in Hz) for this plugin and `datastore_storage_remaining` for the datastore plugin. 

Is it useful to measure what *remains* of the resource (will graphing tools such as Grafana handle that well?), or instead what has been used out of a min/max range of a resource? In other words, if a high value for `cpu_remaining` is given will Grafana consider that a good thing, or a bad thing? You and I can look at the number/label and figure it out, but I wonder if instead the focus should stay on what has been used (`cpu_used` in Hz).

I'm trying to make sure I'm thinking about this right way before I update all of the plugins.

Thanks.

**Reply by Byolock** [reply]
This is what a Graph showing remaining resources can look like in Grafana :

![grafik](https://user-images.githubusercontent.com/25748003/138436633-795d5ad5-3486-4e16-b12a-348e931ed187.png)

This graph shows how much memory is free/available. I opened lots of chrome tabs on this server at the end of the graph so you can actually see how it would look if the available memory gets reduced.

The Orange and Red lines represent the warning and critical values which are configured in my monitoring setup, but Grafana has no understanding of what they mean, its configured as just another value to draw a graph for. Nothing would happen inside Grafana if the available memory would fall below the threshold.

**Grafana doesn't really consider anything good or bad**, there is a threshold feature which does add critical and warning values, but I think it does not matter for your decision because this feature isn't used with Icinga / Nagios. If you want to know more about this threshold feature read the next paragraph, but you are free to skip it I really don't think its relevant.

> 
> For Grafana there is a threshold feature which would be able to consider a value good or bad, but I haven't seen anybody who got it to work in any way together with Icinga2 (and I think any Nagios Version should be the same in this regard). 
> The problem is you can't choose performance data values for the Grafana threshold feature, so you would have to edit every graph with the thresholds you want for this service and it would be impossible to have different thresholds for different hosts. Simply put there is no way to sync the grafana thresholds to any of your Icinga / Nagios configuration.


So for `cpu_remaining` and `datastore_storage_remaining` I would say it comes down to personal preference. If you got any remaining measurement it will mean it's bad when it goes down and the maximum (or minimum, guess it depends on how you look at it) value will be 0. This is probably not intuitive for most people, but it is definitely technically possible to visualize it, as you can see with the image in the beginning of this post. Interpretation of what is good or bad, is entirely up to the human looking at it.

**Reply by atc0005** [reply]
> So for `cpu_remaining` and `datastore_storage_remaining` I would say it comes down to personal preference. If you got any remaining measurement it will mean it's bad when it goes down and the maximum (or minimum, guess it depends on how you look at it) value will be 0. This is probably not intuitive for most people, but it is definitely technically possible to visualize it, as you can see with the image in the beginning of this post. Interpretation of what is good or bad, is entirely up to the human looking at it.

Thanks for the feedback and the sample graph. I'm definitely thinking of the human perspective here as I'd like for the values to make sense (if possible) at a glance.

Continuing to think out loud:

Would it change things if a `*_used` metric were provided for every case where a `*_remaining` metric were provided? This would provide at these metrics:

1. usage (`%`)
1. total
1. used
1. remaining

I see the `%` (1) use case as being one of the more useful high-level metrics with the others helping to track specifics.

Pardon my ignorance, but are you able to select specific metrics in Grafana that you want to track and those that you don't? I'm wondering what the downsides are of supplying all of four of the listed metric types instead of just 1 and 3 (the one you indicated is more intuitive). If the Grafana user is forced into using all supplied metrics, I can see the upside of aggressively limiting what metrics are emitted, but if the metrics can be processed as desired then I'm seeing providing the full set as a useful feature.

Thanks for being a sounding board.

**Reply by Byolock** [reply]
I agree the % (1) metric will definitely be the one to look at primarily. 

Grafana does get it's data from an time based database (influxdb or graphite) and you have to explicitly state which metric you want to show by creating a query to this database. That can look like this :

![grafik](https://user-images.githubusercontent.com/25748003/138473411-ffa2e3c4-09fc-455f-abf8-03180736a6e1.png)
(If you should notice the "OR" in the query is obviously wrong should have been "AND")

- _vmware-cpu_ is the name of the command definition
- _CPU-Last_ is the name of my service definition
- as metric I've choosen _cpu_total_

in the second line at field(value), value means the value of the metric cpu_total, it also does offer the fields (in case of you cpu plugin):

- warn 
- crit
- execution time
- latency
- max
- min
- max_check_attempts

Based on what you choose as the metric and in the field it does show you a line of the chosen value.
You can have as many, or few of these queries in one graph as you like. 

This is a Graph created from data from your cpu plugin with these three queries :

- metric _cpu_usage_ & field _value_
- metric _cpu_usage_ & field _warn_
- metric _cpu_usage_ & field _crit_

![grafik](https://user-images.githubusercontent.com/25748003/138478133-ee218f66-5dc2-4fbd-9a24-3adf5d73912f.png)

tl;dr You can choose which metrics to display and split them in as many graphs as you want. There is no downside in supplying more metrics other than probably a bigger database, it won't ruin the graph of anybody.

**Reply by tubby1981** [reply]
This is interesting. We use nagios for alerting. For graphs we use observium and grafana. It would be nice if we can create a VMware dashboard. Maybe I can also do some testing with it. But their are also other plugins for VMware for building grafana dashboards.

**Reply by atc0005** [reply]
> @Byolock: I agree the % (1) metric will definitely be the one to look at primarily.
>
> ...
> tl;dr You can choose which metrics to display and split them in as many graphs as you want. There is no downside in supplying more metrics other than probably a bigger database, it won't ruin the graph of anybody.

Thanks for the overview, that is helpful. I'll work on expanding the metrics currently offered by plugins updated thus far so that all four of the mentioned fields are available. If you spot any metrics that I've missed or have feedback in general to share, please do so. Prior to this discussion thread, I've not worked with Nagios performance data, so this has been a good learning experience.

> @tubby1981: This is interesting. We use nagios for alerting. For graphs we use observium and grafana. 

Thanks for the mention of observium. I'm not sure I've looked at it before, I'll look into it.

**Reply by atc0005** [reply]
> I'll work on expanding the metrics currently offered by plugins updated thus far so that all four of the mentioned fields are available.

Added in GH-438.

**Comment by atc0005** [comment]
@Byolock The initial implementation of Performance Data metrics is complete. Going to tag a release candidate soon.

#480 was opened to cover further tweaks to the `*_remaining` metrics. At this point, I'm considering removing the `Min` and `Max` values in an effort to remove any "gotchas" associated with understanding the purpose of the metric. 

If we flip `Min` & `Max` ( `*_remaining` metrics) to make the results easier to understand, I wonder whether it provides any real contrast against the `*_used` metrics. I don't have a lot of experience with tools such as Grafana, but I'm guessing that they already provide some means of easily providing an inverse of `*_used` based on the value provided by that metric.

**Reply by atc0005** [reply]
> Going to tag a release candidate soon.

v0.26.0-rc.1 was released a few days ago and v0.26.0 was released today (only changes were CHANGELOG updates).

The bulk of the work for implementing performance data metrics is now complete.

I plan to revisit #480 sometime in the days ahead after I give it some more thought and after any potential feedback is provided.

**Comment by atc0005** [comment]
v0.28.0-rc1 contains changes that wrap-up the work for this request.

Considering this complete.
